port/src/devcons.rs

```diff
@@ -0,0 +1,40 @@
+use crate::mcslock::{Lock, LockNode};
+
+const fn ctrl(b: u8) -> u8 {
+    b - b'@'
+}
+
+#[allow(dead_code)]
+const BACKSPACE: u8 = ctrl(b'H');
+#[allow(dead_code)]
+const DELETE: u8 = 0x7F;
+#[allow(dead_code)]
+const CTLD: u8 = ctrl(b'D');
+#[allow(dead_code)]
+const CTLP: u8 = ctrl(b'P');
+#[allow(dead_code)]
+const CTLU: u8 = ctrl(b'U');
+
+pub trait ConsoleDriver {
```

Comment:
I would name this `Console`, and put most of the console code here, and then have a trait called `Uart` (or something) that implemented that implemented the `putb` function that `Console` is written in terms of. `Console` would be parameterized with a `&'static mut Uart`. That is, instead of have a `ConsoleDriver` trait that knows about writing a byte to a UART, you'd have a `Console` struct and a Uart trait.

Reply:
So far I'm looking at something like this to create a global console:
```
static mut UART: Uart16550 = Uart16550 { port: 0x3f8 };
static mut CONSOLE: Console = Console { uart: &mut UART };
```

Rust doesn't like this.  I suspect I need to use something like lazy_static or once_cell.  Any preference?

Reply:
Yes. It'll complain about the static mut ref in Console.

Don't bother with lazy_static (which is buggy) or once_cell (which really only gives you a OnceBool for no_std code). Instead, consider making Console take the Uart by value and assume ownership of it; then simply create in the Console constructor.